File: internal/agentidentity/agentidentity_test.go
package agentidentity

import "testing"

// Tabla canónica de resolución: MISMA tabla byte-por-byte que
// scripts/lib/agent-identity.test.js (paridad E1, REQ-agent-identity-003).
// El caso `plugin-host:sdd-spec` es la REGRESIÓN del bug de prefijo de host
// que la igualdad estricta `sdd-${phase}` no reconocía.
var resolutionCases = []struct {
	name     string
	raw      string
	expected string
}{
	{"unprefixed sdd identity", "sdd-spec", "sdd-spec"},
	{"host prefix strip (regresion prefijo)", "plugin-host:sdd-spec", "sdd-spec"},
	{"review prefix strip", "host:review-runtime", "review-runtime"},
	{"trim whitespace", "  sdd-spec  ", "sdd-spec"},
	{"double prefix fails closed", "a:b:sdd-spec", Unresolved},
	{"empty prefix fails closed", ":sdd-spec", Unresolved},
	{"empty suffix fails closed", "host:", Unresolved},
	{"empty sdd suffix", "sdd-", Unresolved},
	{"empty string", "", Unresolved},
	{"blank string", "   ", Unresolved},
	{"foreign review name", "review-invented", Unresolved},
	{"unlisted review name", "review-reliability", Unresolved},
	{"bare sdd without suffix", "sdd", Unresolved},
	{"uppercase is foreign", "SDD-spec", Unresolved},
}

func TestResolveCanonicalAgentTable(t *testing.T) {
	for _, tc := range resolutionCases {
		if got := ResolveCanonicalAgent(tc.raw); got != tc.expected {
			t.Errorf("ResolveCanonicalAgent(%q) = %q, want %q", tc.raw, got, tc.expected)
		}
	}
}

func TestResolveCanonicalAgentReviewAllowlist(t *testing.T) {
	if len(ReviewAgents) != 6 {
		t.Fatalf("ReviewAgents: got %d entries, want 6", len(ReviewAgents))
	}
	for _, agent := range ReviewAgents {
		if got := ResolveCanonicalAgent(agent); got != agent {
			t.Errorf("ResolveCanonicalAgent(%q) = %q, want identity", agent, got)
		}
		if got := ResolveCanonicalAgent("host:" + agent); got != agent {
			t.Errorf("ResolveCanonicalAgent(host:%s) = %q, want %q", agent, got, agent)
		}
	}
}

var phaseKeyCases = []struct {
	canonical string
	expected  string
}{
	{"sdd-spec", "spec"},
	{"sdd-apply", "apply"},
	{"review-runtime", "review-runtime"},
	{"review-correction", "review-correction"},
	{"unknown-agent", ""},
	{Unresolved, ""},
}

func TestDerivePhaseKeyTable(t *testing.T) {
	for _, tc := range phaseKeyCases {
		if got := DerivePhaseKey(tc.canonical); got != tc.expected {
			t.Errorf("DerivePhaseKey(%q) = %q, want %q", tc.canonical, got, tc.expected)
		}
	}
}

// legacyEmitterOutput replica la lógica que hoy emiten los hooks para nombres
// sin prefijo (O1): strip `sdd-` / review self / "".
func legacyEmitterOutput(agentName string) string {
	review := map[string]bool{
		"review-change": true, "review-trust": true, "review-runtime": true,
		"review-evolution": true, "review-efficiency": true, "review-correction": true,
	}
	if len(agentName) >= 4 && agentName[:4] == "sdd-" {
		return agentName[4:]
	}
	if review[agentName] {
		return agentName
	}
	return ""
}

func TestO1CompatibilityUnprefixedNames(t *testing.T) {
	unprefixed := []string{
		"sdd-spec", "sdd-apply", "sdd-verify", "sdd-document", "sdd-orchestrator",
		"review-change", "review-trust", "review-runtime",
		"review-evolution", "review-efficiency", "review-correction",
	}
	if len(unprefixed) == 0 {
		t.Fatal("unprefixed set must be non-empty")
	}
	for _, name := range unprefixed {
		canonical := ResolveCanonicalAgent(name)
		if canonical != name {
			t.Errorf("identity: ResolveCanonicalAgent(%q) = %q", name, canonical)
		}
		if got := DerivePhaseKey(canonical); got != legacyEmitterOutput(name) {
			t.Errorf("phase key: DerivePhaseKey(%q) = %q, want legacy %q", name, got, legacyEmitterOutput(name))
		}
	}
}

// paritySet: set representativo del spec; DEBE afirmar el mismo mapa esperado
// que la tabla Parity E1 en scripts/lib/agent-identity.test.js.
var paritySet = []struct {
	raw               string
	expectedCanonical string
	expectedKey       string
}{
	{"sdd-spec", "sdd-spec", "spec"},
	{"host:sdd-spec", "sdd-spec", "spec"},
	{"review-runtime", "review-runtime", "review-runtime"},
	{"host:review-runtime", "review-runtime", "review-runtime"},
	{"review-invented", Unresolved, ""},
	{"plugin-host:sdd-spec", "sdd-spec", "spec"},
}

func TestParityE1RepresentativeNames(t *testing.T) {
	if len(paritySet) < 6 {
		t.Fatal("parity set must cover the representative spec inputs")
	}
	for _, tc := range paritySet {
		canonical := ResolveCanonicalAgent(tc.raw)
		if canonical != tc.expectedCanonical {
			t.Errorf("canonical: ResolveCanonicalAgent(%q) = %q, want %q", tc.raw, canonical, tc.expectedCanonical)
		}
		key := ""
		if canonical != Unresolved {
			key = DerivePhaseKey(canonical)
		}
		if key != tc.expectedKey {
			t.Errorf("phase key: %q -> %q, want %q", tc.raw, key, tc.expectedKey)
		}
	}
}

File: scripts/lib/agent-identity.test.js
"use strict";

const test = require("node:test");
const assert = require("node:assert/strict");

const {
  UNRESOLVED,
  REVIEW_AGENTS,
  resolveCanonicalAgent,
  derivePhaseKey,
} = require("./agent-identity.js");

// Tabla canónica de resolución (REQ-agent-identity-001). El caso
// `plugin-host:sdd-spec` es la REGRESIÓN del bug de prefijo de host que la
// igualdad estricta `sdd-${phase}` no reconocía.
const RESOLUTION_CASES = [
  // unprefixed sdd identity
  ["sdd-spec", "sdd-spec"],
  // host/plugin prefix strip (regresión prefijo)
  ["plugin-host:sdd-spec", "sdd-spec"],
  ["host:review-runtime", "review-runtime"],
  // trim
  ["  sdd-spec  ", "sdd-spec"],
  // double prefix / malformed colon → fail closed
  ["a:b:sdd-spec", UNRESOLVED],
  [":sdd-spec", UNRESOLVED],
  ["host:", UNRESOLVED],
  // empty suffix after sdd-
  ["sdd-", UNRESOLVED],
  // empty / non-string
  ["", UNRESOLVED],
  ["   ", UNRESOLVED],
  [undefined, UNRESOLVED],
  [null, UNRESOLVED],
  [42, UNRESOLVED],
  // closed-world review set: foreign review-* names fail
  ["review-invented", UNRESOLVED],
  ["review-reliability", UNRESOLVED],
  ["sdd", UNRESOLVED],
  ["SDD-spec", UNRESOLVED],
];

for (const [raw, expected] of RESOLUTION_CASES) {
  test(`resolveCanonicalAgent: ${JSON.stringify(raw)} -> ${expected}`, () => {
    assert.equal(resolveCanonicalAgent(raw), expected);
  });
}

for (const agent of REVIEW_AGENTS) {
  test(`resolveCanonicalAgent: allowlisted review agent ${agent} resolves to itself`, () => {
    assert.equal(resolveCanonicalAgent(agent), agent);
    assert.equal(resolveCanonicalAgent(`host:${agent}`), agent);
  });
}

// --- derivePhaseKey ----------------------------------------------------------

const PHASE_KEY_CASES = [
  ["sdd-spec", "spec"],
  ["sdd-apply", "apply"],
  ["review-runtime", "review-runtime"],
  ["review-correction", "review-correction"],
  ["unknown-agent", ""],
  [UNRESOLVED, ""],
];

for (const [canonical, expected] of PHASE_KEY_CASES) {
  test(`derivePhaseKey: ${canonical} -> ${JSON.stringify(expected)}`, () => {
    assert.equal(derivePhaseKey(canonical), expected);
  });
}

// --- O1 compatibilidad (REQ-agent-identity-002) ------------------------------

// Réplica literal de la lógica que hoy emiten los hooks para nombres sin
// prefijo: strip `sdd-` / review self / "". La resolución canónica debe
// producir exactamente estos valores para todo nombre válido sin prefijo.
function legacyEmitterOutput(agentName) {
  if (agentName.startsWith("sdd-")) return agentName.slice("sdd-".length);
  return new Set(["review-change", "review-trust", "review-runtime", "review-evolution", "review-efficiency", "review-correction"]).has(agentName)
    ? agentName
    : "";
}

test("O1 compat: todo nombre sin prefijo produce identidad + phase key idénticos a los emitidos hoy", () => {
  const unprefixed = [
    "sdd-spec",
    "sdd-apply",
    "sdd-verify",
    "sdd-document",
    "sdd-orchestrator",
    ...REVIEW_AGENTS,
  ];
  assert.ok(unprefixed.length > 0);
  for (const name of unprefixed) {
    const canonical = resolveCanonicalAgent(name);
    assert.equal(canonical, name, `identidad: ${name}`);
    assert.equal(derivePhaseKey(canonical), legacyEmitterOutput(name), `phase key: ${name}`);
  }
});

// --- Paridad E1 (REQ-agent-identity-003) -------------------------------------

// Set representativo del spec; el espejo Go (internal/agentidentity) afirma el
// MISMO mapa esperado. Cualquier divergencia entre runtimes rompe esta tabla.
const PARITY_SET = [
  ["sdd-spec", "sdd-spec", "spec"],
  ["host:sdd-spec", "sdd-spec", "spec"],
  ["review-runtime", "review-runtime", "review-runtime"],
  ["host:review-runtime", "review-runtime", "review-runtime"],
  ["review-invented", UNRESOLVED, ""],
  // caso de regresión: nombre prefijado que la igualdad estricta rechazaba
  ["plugin-host:sdd-spec", "sdd-spec", "spec"],
];

test("Paridad E1: resultados esperados idénticos a los del espejo Go", () => {
  assert.ok(PARITY_SET.length >= 6);
  for (const [raw, expectedCanonical, expectedKey] of PARITY_SET) {
    const canonical = resolveCanonicalAgent(raw);
    assert.equal(canonical, expectedCanonical, `canonical: ${raw}`);
    assert.equal(
      canonical === UNRESOLVED ? "" : derivePhaseKey(canonical),
      expectedKey,
      `phase key: ${raw}`,
    );
  }
});
